File: server/wwwroot/js/Interop/Ws.js

import { Record } from "../fable_modules/fable-library-js.5.13.0/Types.js";
import { obj_type, record_type, string_type, array_type, uint8_type, lambda_type, unit_type } from "../fable_modules/fable-library-js.5.13.0/Reflection.js";

export class Handlers extends Record {
    constructor(OnOpen, OnClose, OnBinary, OnText) {
        super();
        this.OnOpen = OnOpen;
        this.OnClose = OnClose;
        this.OnBinary = OnBinary;
        this.OnText = OnText;
    }
}

export function Handlers_$reflection() {
    return record_type("Client.Interop.Ws.Handlers", [], Handlers, () => [["OnOpen", lambda_type(unit_type, unit_type)], ["OnClose", lambda_type(unit_type, unit_type)], ["OnBinary", lambda_type(array_type(uint8_type), unit_type)], ["OnText", lambda_type(string_type, unit_type)]]);
}

export class Connection extends Record {
    constructor(Socket) {
        super();
        this.Socket = Socket;
    }
}

export function Connection_$reflection() {
    return record_type("Client.Interop.Ws.Connection", [], Connection, () => [["Socket", obj_type]]);
}

export function connect(url, handlers) {
    const socket = new WebSocket(url);
    socket.binaryType = "arraybuffer";
    socket.onopen = ((_arg) => {
        handlers.OnOpen();
    });
    socket.onclose = ((_arg_1) => {
        handlers.OnClose();
    });
    socket.onmessage = ((ev) => {
        const data = ev.data;
        if (typeof data === "string") {
            const s = data;
            handlers.OnText(s);
        }
        else {
            handlers.OnBinary(new Uint8Array(data));
        }
    });
    return new Connection(socket);
}

export function isOpen(conn) {
    return conn.Socket.readyState === 1;
}

export function sendBinary(conn, bytes) {
    if (isOpen(conn)) {
        conn.Socket.send(bytes);
    }
}

export function sendText(conn, text) {
    if (isOpen(conn)) {
        conn.Socket.send(text);
    }
}

File: client/Interop/Ws.fs
/// Minimal WebSocket wrapper. Binary frames carry Yjs update bytes, text frames
/// carry JSON awareness messages - see server/Rooms.fs and Doc.fs. Kept dynamic
/// (same rationale as Canvas.fs) instead of fighting typed DOM bindings.
module Client.Interop.Ws

open Fable.Core
open Fable.Core.JsInterop

/// Reference to the browser's global `WebSocket` constructor. The explicit name is
/// required - without it Fable uses this binding's own F# name as the global lookup,
/// which would (wrongly) look for a global literally called `WebSocketCtor`.
[<Global("WebSocket")>]
let private WebSocketCtor: obj = jsNative

type Handlers =
    { OnOpen: unit -> unit
      OnClose: unit -> unit
      OnBinary: byte[] -> unit
      OnText: string -> unit }

type Connection = { Socket: obj }

let connect (url: string) (handlers: Handlers) : Connection =
    let socket: obj = createNew WebSocketCtor url
    socket?binaryType <- "arraybuffer"
    socket?onopen <- fun (_: obj) -> handlers.OnOpen()
    socket?onclose <- fun (_: obj) -> handlers.OnClose()

    socket?onmessage <-
        fun (ev: obj) ->
            let data: obj = ev?data

            match data with
            | :? string as s -> handlers.OnText s
            | _ ->
                // binaryType = "arraybuffer", so anything non-string is an ArrayBuffer.
                let bytes: byte[] = JS.Constructors.Uint8Array.Create(data) |> unbox
                handlers.OnBinary bytes

    { Socket = socket }

let isOpen (conn: Connection) : bool =
    let state: float = conn.Socket?readyState
    state = 1.0 // WebSocket.OPEN

let sendBinary (conn: Connection) (bytes: byte[]) : unit =
    if isOpen conn then
        conn.Socket?send (bytes)

let sendText (conn: Connection) (text: string) : unit =
    if isOpen conn then
        conn.Socket?send (text)
